serviceregistration/consul/consul_service_registration.go

```diff
@@ -212,10 +211,21 @@ func NewConsulServiceRegistration(conf map[string]string, logger log.Logger) (sr
 		checkTimeout:        checkTimeout,
 		disableRegistration: disableRegistration,
 
-		notifyActiveCh:      make(chan notifyEvent),
-		notifySealedCh:      make(chan notifyEvent),
-		notifyPerfStandbyCh: make(chan notifyEvent),
+		notifyActiveCh:      make(chan bool),
+		notifySealedCh:      make(chan bool),
+		notifyPerfStandbyCh: make(chan bool),
+
+		isActive:      state.IsActive,
+		isSealed:      state.IsSealed,
+		isPerfStandby: state.IsPerformanceStandby,
 	}
+	go func() {
+		if err := c.runServiceRegistration(&sync.WaitGroup{}, shutdownCh, redirectAddr); err != nil {
```

Comment:
Previously this waitgroup was passed into the run function via server.go. The caller can no longer wait for these go routines to close since we are creating it here. Consider allowing it to be passed as a parameter if you adopt the Run/Stop API

Reply:
In testing, I found that it wasn't necessary. When triggering off the `shutdownCh` to execute cleanup, I found that all the commands were fully executed before everything ended. I was concerned about using the waitgroup because it added more complexity to the implementation. I think a lot of folks aren't practiced with waitgroups, and I've actually seen a lot of bugs from folks using them, having incorrect logic around calling done, and applications hanging forever on shutdown. 

So, I'd be in favor of either avoiding passing it in, or, if we implement an `OnStop` method, we could alternatively write a wrapper who handles waiting until `OnStop` is complete, using the waitgroup if necessary. Then we just use the waitgroup once and we don't risk introducing a bug with shutdown behavior whenever we add a new service registration.

Reply:
Waitgroup was added here: https://github.com/hashicorp/vault/pull/8123/commits/7070a6a8451bcfa37387e57a2f2f9d39ee5bc34d